Userland/Libraries/LibGUI/VimEditingEngine.cpp

```diff
@@ -63,8 +63,111 @@ bool VimEditingEngine::on_key_in_insert_mode(const KeyEvent& event)
     return false;
 }
 
+char VimEditingEngine::numeric_key_value(const KeyCode key)
+{
+    switch (key) {
+    case (KeyCode::Key_0):
+        return '0';
+    case (KeyCode::Key_1):
+        return '1';
+    case (KeyCode::Key_2):
+        return '2';
+    case (KeyCode::Key_3):
+        return '3';
+    case (KeyCode::Key_4):
+        return '4';
+    case (KeyCode::Key_5):
+        return '5';
+    case (KeyCode::Key_6):
+        return '6';
+    case (KeyCode::Key_7):
+        return '7';
+    case (KeyCode::Key_8):
+        return '8';
+    case (KeyCode::Key_9):
+        return '9';
+    default:
+        return '\0';
+    }
+}
+
+void VimEditingEngine::clear_command_multiplier()
+{
+    m_command_multiplier.clear();
+}
+
+void VimEditingEngine::vim_move_down(const KeyEvent& event)
+{
+    // FIXME: probably not an efficient way to do this
+    if (m_command_multiplier.is_empty()) {
+        move_one_down(event);
+    } else {
+        String multiplier_string = m_command_multiplier.build();
+        uint32_t multiplier = multiplier_string.to_uint().value();
+        for (uint32_t i = 0; i < multiplier; ++i) {
+            move_one_down(event);
+        }
+        clear_command_multiplier();
+    }
+}
+
+void VimEditingEngine::vim_move_up(const KeyEvent& event)
+{
+    if (m_command_multiplier.is_empty()) {
+        move_one_up(event);
+    } else {
+        String multiplier_string = m_command_multiplier.build();
+        uint32_t multiplier = multiplier_string.to_uint().value();
+        for (uint32_t i = 0; i < multiplier; ++i) {
+            move_one_up(event);
+        }
+        clear_command_multiplier();
+    }
+}
+
+void VimEditingEngine::vim_move_left(const KeyEvent& event)
+{
+    if (m_command_multiplier.is_empty()) {
+        move_one_left(event);
+    } else {
+        String multiplier_string = m_command_multiplier.build();
+        uint32_t multiplier = multiplier_string.to_uint().value();
+        for (uint32_t i = 0; i < multiplier; ++i) {
+            move_one_left(event);
+        }
+        clear_command_multiplier();
+    }
+}
+
+void VimEditingEngine::vim_move_right(const KeyEvent& event)
+{
+    if (m_command_multiplier.is_empty()) {
+        move_one_right(event);
+    } else {
+        String multiplier_string = m_command_multiplier.build();
+        uint32_t multiplier = multiplier_string.to_uint().value();
+        for (uint32_t i = 0; i < multiplier; ++i) {
+            move_one_right(event);
+        }
+        clear_command_multiplier();
+    }
+}
+
+void VimEditingEngine::add_to_command_multiplier(char value)
+{
+    // FIXME: there are more efficient ways to do this
+    if (m_command_multiplier.length() < 5) {
+        m_command_multiplier.append(value);
+    }
+}
+
 bool VimEditingEngine::on_key_in_normal_mode(const KeyEvent& event)
 {
+    char numeric_value = numeric_key_value(event.key());
+    if (numeric_value != '\0') {
+        add_to_command_multiplier(numeric_value);
+        return true;
+    }
```

Comment:
`on_key_in_normal_mode` is already really chonky, but it feels like the specific implementation you're going for, moving N times in a direction, only applies 1) if the previous key was not d (in that case we should be looking for more input for "what" to `delete`) and  2) if there are no modifiers pressed.

so, I'd recommend moving this check to the !ctrl, !alt and !shift check that starts on line 337 below. Otherwise if you do, `5dd5` or something you'll store `55` in the m_command_multiplier instead of `5`.

Reply:
actually maybe i'm wrong about the recommendation. But the m_previous_key logic might screw you up if you don't clear the command multiplier in more situations.

Reply:
Good note! You're absolutely right that I got those things wrong, I'll start by taking your recommendation and moving it to the no-modifier part of the code and then think of some scenarios when the multiplier needs to be cleared and when it is okay to add to it (and capture the key) :)